proxy/http/HttpTransact.cc

```diff
@@ -4464,7 +4464,7 @@ HttpTransact::handle_cache_operation_on_forward_server_response(State *s)
     //   before issuing a 304
     if (s->cache_info.action == CACHE_DO_WRITE || s->cache_info.action == CACHE_DO_NO_ACTION ||
         s->cache_info.action == CACHE_DO_REPLACE) {
-      if (s->is_cacheable_and_negative_caching_is_enabled) {
+      if (s->is_cacheable_and_negative_caching_is_enabled && is_negative_caching_appropriate(s)) {
```

Comment:
I worry about `s->is_cacheable_and_negative_caching_is_enabled` is not set correctly in line 4405. Prior to the 8eb68266167d8f8b3fa3a00ca9f6b7889e8ec101, when this function set `s->is_cacheable_and_negative_caching_is_enabled` (renamed from `s-> negative_caching`), it called `is_negative_caching_appropriate(s)`. Should we roll it back? 
```
4405 -      s->negative_caching = is_negative_caching_appropriate(s) && cacheable;
4405 +      s->is_cacheable_and_negative_caching_is_enabled = cacheable && s->txn_conf->negative_caching_enabled;
```
https://github.com/apache/trafficserver/commit/8eb68266167d8f8b3fa3a00ca9f6b7889e8ec101#diff-7f286e5169bc2d28c13f1ceaf3bb482b36cdb0743fbac4040fd4cd5a9c0897c1L4405-R4405

/cc @bneradt

Reply:
FWIW, this change fixed a leak which I reported on https://github.com/apache/trafficserver/issues/7380#issuecomment-748701998.

Reply:
In case it's helpful, I did these suggested changes on the following branch:
https://github.com/bneradt/trafficserver/tree/fix_negative_caching_leak

@masaori335 and @shinya1020 : can you please try things out with this branch to see whether it addresses the header and leak issues? If so, we can update this PR with that change. Or I can create a new PR if you prefer. Just let me know.

Thanks,
Brian

Reply:
I verified the suggested change fixes the leak 👍

Reply:
Great! Thank you @masaori335 for the quick confirmation. I also ran the negative-caching AuTest with this change on master and verified that the original negative caching fix still works as expected.

@shinya1020 : Can you please try out this change on your branch and verify it addresses the Expires header issue you were seeing? If so, I suggest we update this PR and get this merged in.

Reply:
@bneradt Thank you for your suggestions. I confirmed that Expires header issue no longer occurs on your branch.

Reply:
I've updated this PR to @bneradt patch.